### PR TITLE
feat: support DML with Returning clause in Connection API

### PR DESCRIPTION
**Pull request by rajatbhatta**
This PR adds support for running DML statements with Returning clause in the Connection API, and incorporates the following:
- API documentation changes to document the behaviour of `execute`, `executeAsync`, `executeQuery`, `executeQueryAsync`, `executeUpdate`, `executeUpdateAsync`, `executeBatchUpdate`, and `executeBatchUpdateAsync` APIs when executing DML statements with Returning clause.
- Parses the SQL string to check if the DML statement has a Returning clause. Parsing logic has been added for both GoogleStandardSql and PostgreSql dialects.
- Changes to MockSpannerServiceImpl to put results for DML statements with Returning clause, and to add support in `executeBatchDml`.
- Integration tests for running DML statements with Returning clause using each of the above APIs, for both GoogleStandardSql and PostgreSql dialects.
- Unit tests for abort scenarios while running DML statements with Returning clause.
- Parser tests for both GoogleStandardSql and PostgreSql dialects.

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ReadWriteTransaction.java
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ReadWriteTransaction.java:345]: Can we add a couple of tests to https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java for DML statements with `RETURNING` clauses to ensure that it works correctly with the internal retry mechanism of the Connection API?

Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ReadWriteTransaction.java:345]: Added the tests.

**Comment by rajatbhatta** [comment]
Note: Integration tests are not expected to pass until DML with Returning clause support is launched.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:276]: Also here: I think this could be made more efficient by at least removing the `compile` statement out of this method, but I also think that you don't need a regex at all. You can just do a simple string check if you first replace all double spaces with single spaces, and make the SQL string all lower or upper case.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java:61]: nit: add a TODO here + a Do not merge label in the PR itself to make sure this is modified before the PR is merged.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java:1138]: Note that while this works fine here and is semantically correct, it might be a little confusing, because (I think) the backend does not allow DML statements with a RETURNING clause to be executed using `BatchDml`.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java:160]: I think it would be better to leave these tests as they were and add separate tests for DML statements with returning clauses. This potentially removes existing test coverage for DML statements without a returning clause, and those are handled differently than DML statements with a returning clause, so we should have separate tests for each.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java:320]: An interesting question here is whether the client should already throw an exception when you try to execute it, or whether we should rely on the backend doing so (at least, as far as I can remember, `BatchDml` does not support returning clauses).

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbstractMockServerTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbstractMockServerTest.java:106]: This doesn't seem right. A DML statement with a `RETURNING` clause would return a result set with at least one column.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1316]: Could we also add some tests with (string) literals surrounding the `then return` clause?

So for example `delete from x where y='then return'` and `delete from x where y='z'then return *`

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1395]: Same as above: Can we add some tests with string literals?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1383]: We also need tests with nested comments (see also my comment on the parser itself)

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java:187]: nit: use `assertThrows` instead of a `try-catch-fail` construct

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java:174]: Remove or replace with an actual logger

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java
Comment (olavloite) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java:213]: If this actually works, then my previous comments regarding `BatchDml` and returning are void. But I also think you should check with the developers of this feature whether this is intentional, as the result of the `RETURNING` clause is lost in these cases.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/Connection.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/Connection.java:991]: DML with Returning clause over `BatchDML` is allowed. We can find corresponding integration tests for `BatchDml` and `BatchDmlAsync` available in ITDmlReturningTest in this PR.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/Connection.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/Connection.java:1014]: DML with Returning clause over BatchDML is allowed. We can find corresponding integration tests for BatchDml and BatchDmlAsync available in ITDmlReturningTest in this PR.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java:213]: Yes, this feature is intentional, as discussed with the developers of this feature.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:317]: I had the exception at the end of this function for unclosed literal and I realise that it is redundant as we already have this handling when we remove comments from string. I'll remove that handling, and return true directly from here.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java:117]: I think this would not go wrong. We are checking for space before/after only for the outermost comment. I've added a test case having nested comments too in that regard.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1393]: I've already added test with nested comments. Responded on the above comment.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java]: Thanks for the points!

- I have now optimised the isReturning method to only look for a particular length of substring (for example - in GoogleStandardSql, we would need to only check for substrings of length 13 and check for any of the characters out of ``)`"'`` before `then return` and any of the characters out of ``(`"'`` after it). Regex check would then have fairly same time complexity as normal comparison.
- I've now Incorporated point 1 as well.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java]: Added answer to this in https://github.com/googleapis/java-spanner/pull/1978#discussion_r952676785.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/MockSpannerServiceImpl.java:1138]: DML with Returning clause over BatchDML is allowed. Have added corresponding integration tests for `executeBatchDml` and `executeBatchDmlAsync` in ITDmlReturningTest in this PR.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/it/ITDmlReturningTest.java]: Removed it.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java:320]: DML with Returning clause over `BatchDML` is allowed. Have added corresponding integration tests for `executeBatchDml` and `executeBatchDmlAsync` in ITDmlReturningTest in this PR.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1316]: Added.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbstractMockServerTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbstractMockServerTest.java]: Modified it to set row type to a StructType with 1 field (so that result set has 1 column).

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/AbortedTest.java]: Added separate tests for DML statements with returning clause.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java:318]: This is still a very expensive check, as we always iterate through the entire SQL string and repeatedly check for a regex match. Can we improve this a little by doing the following:
1. Do a pre-check to see if the (lower)-case string contains `"returning"` at all. If it does not, we know for sure that the regex will also never match. See https://gist.github.com/olavloite/a857ff860c4f12c17d51712bed57b9d3 for a small benchmark program. Some quick tests indicate that having such a pre-check improves performance with a factor 10x (400ms for 100,000 checks as opposed to 4,000ms on my local machine).
2. Do the while loop in the opposite order (i.e. start at `index=sql.length()`). If we do a pre-check and that check returns true, the probability that we will find a match is a lot bigger, as `returning` is not something that is very common in any given SQL string. And a `returning` clause will be somewhere at the end of the SQL string.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java
Comment (olavloite) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:283]: Could we also do a couple of optimizations here:
1. Do a `sql.contains("return") && sql.contains("then")` check before anything else, as that would allow us to return false quickly in most cases.
2. (Maybe in a separate PR, as this is more complicated here because of the checking for escaping etc.) Loop in the opposite direction if the quick check succeeds?

Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:283]: Good point! I'll work on enhancement #2 in a separate PR, and will incorporate enhancement #1 in this PR for now.

Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:283]: The check can be `sql.contains("then return")` as we must have already replaced all consecutive whitespaces with a single space.

Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:283]: No, I would recommend that we do not do the last. So the order of operations would roughly be:
1. sql.toLowerCase()
2. sql.contains("return") && sql.contains("then")
3. sql.replaceAll("\\s+", " ")

The reason for this is that the `replaceAll` is also quite expensive, as it is also a regular expression check for each element in the string.

Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:283]: Makes sense.

Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java:283]: Made this change.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (c2nes) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1366]: PostgreSQL supports some additional syntaxes for strings, [Dollar-Quoted String Constants](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-DOLLAR-QUOTING) in particular, that you may want to add test cases for. For instance,

```sql
UPDATE foo SET bar = $$RETURNING a, b, c$$ WHERE baz = 123
```

does not have a returning clause. Nor does,


```sql
UPDATE foo SET bar = $foobar$RETURNING a, b, c$foobar$ WHERE baz = 123
```

nor,

```sql
UPDATE foo SET bar = $returning$Hello, world!$returning$ WHERE baz = 123
```

Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1366]: @c2nes: Thanks for the test cases! Added these as well in https://github.com/googleapis/java-spanner/pull/1978/commits/9f59715e76511dceb43825a66c06f290978ed02e.

**Review by olavloite — APPROVED** [review]
@rajatbhatta Thanks for your patience with me for this PR. LGTM.

**Comment by rajatbhatta** [comment]
@olavloite: Thank you for a very detailed review of this PR. :)

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (bvandiver) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1366]: bvandiver=# insert into t1 select 2returning *;
 a 
---
 2
(1 row)

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/AbstractStatementParser.java
Comment (thiagotnunes) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/AbstractStatementParser.java:554]: Can this be `protected`?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java
Comment (thiagotnunes) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java:313]: nit: create a constant for "returning"

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java
Comment (thiagotnunes) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java:64]: It seems we forgot to clean this file?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestEnv.java
Comment (thiagotnunes) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestEnv.java:94]: It seems we forgot to clean this file?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/AbstractStatementParser.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/AbstractStatementParser.java]: Makes sense. Done.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/GceTestEnvConfig.java:64]: I'll clean it before merging. Have added a TODO here.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestEnv.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestEnv.java:94]: I'll clean it before merging. Have added a TODO here.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java
Comment (rajatbhatta) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/StatementParserTest.java:1370]: Thanks for this test case. I've handled this scenario by checking if the substring containing returning doesn't start with a valid first character of an identifier. In that case, the substring can be divided further into substrings. Have added more tests around that as well.

**Comment by rajatbhatta** [comment]
@olavloite: Can you take a quick look at the changes in https://github.com/googleapis/java-spanner/pull/1978/commits/e2a463e4f91c602cd4a237634f557bcf4793de21? Thanks!

**Review by thiagotnunes — APPROVED** [review]

**Comment by olavloite** [comment]
> @olavloite: Can you take a quick look at the changes in [e2a463e](https://github.com/googleapis/java-spanner/commit/e2a463e4f91c602cd4a237634f557bcf4793de21)? Thanks!

Can we also add a test for `insert into t1 select 10.returning*` (this is equivalent with `insert into t1 select 10.0 returning *`)

**Comment by rajatbhatta** [comment]
> > @olavloite: Can you take a quick look at the changes in [e2a463e](https://github.com/googleapis/java-spanner/commit/e2a463e4f91c602cd4a237634f557bcf4793de21)? Thanks!
> 
> Can we also add a test for `insert into t1 select 10.returning*` (this is equivalent with `insert into t1 select 10.0 returning *`)

Added in https://github.com/googleapis/java-spanner/pull/1978/commits/60fcb4cdfb47fdc64924bff914e35d5cd454aa3d.

**Comment by olavloite** [comment]
> 60fcb4c

Thanks. LGTM (once CI errors have been cleared)

**Pull request merged**
